File: crates/config-engine/src/user_config/rule_resource_catalog.rs
//! 规则资源库（catalog）内置精选清单 —— **Rust 是单一真值（SoT）**。
//!
//! 前端曾持有同一张表（`ui/src/shared/rule-resource-catalog.ts` 的 `RULE_RESOURCE_CATALOG`），
//! 现经 `rule_resources_get_catalog` command 下发（常量表 → Rust SoT + 一次 invoke 拉取，
//! 判据见 ~/docs/polaris/design/polaris-dialog-layer-and-governance.md §3.1 Q3）。
//!
//! **本表不再是独立表，而是 `builtin_geo_rulesets()` 的投影**。
//!
//! 此前两张表各自维护、只有交集（33 条精选 vs 28 条随包，重合 19 条），于是「内置」tab 里同时躺着
//! 两类东西：随包出厂的（勾上、不用下）和只是列在这儿的（点了要联网下）。同一个词「内置」指两件事，
//! 用户看不出区别 —— 而列而不随包的那 14 条无一被代码引用（既不在地区分流基线，也不在内置应用预设），
//! 选中它们只会走到 `route.rs` 的 fail-closed 剪枝：规则静默失效。
//!
//! 裁定（2026-07-30）：**内置 = 随包 = 有消费点**，三者恒等。恒等由「投影」而非「一条对拍测试」
//! 保证 —— 测试只能在漂了之后转红，投影让它压根漂不出来。
//!
//! 全量可下载清单仍在「外置」tab（`rule_resources_refresh_catalog` 拉 meta-rules-dat 全量），
//! 想要 `geoip-jp` / `geosite-apple` 这类未随包资源的用户走那条路，不受本次收敛影响。

#![forbid(unsafe_code)]

use super::builtin_geo_rulesets::{builtin_geo_rulesets, BuiltinGeoRuleSet, GeoCategory};
use serde::Serialize;

/// meta-rules-dat `sing` 分支 raw 基址。上游 `MRD_RAW_BASE`。
///
/// 注：`builtin_geo_rulesets::mrd_geo_raw_base()` 是**另一个**基址（多一段 `/geo`），供随包 geo 更新用；
/// 本表的 `path` 已含 `geo/` 前缀，故基址到 `sing/` 为止。二者不可互换。
pub const MRD_RAW_BASE: &str = "https://raw.githubusercontent.com/MetaCubeX/meta-rules-dat/sing/";

/// 仓库内路径（如 `geo/geosite/youtube.srs`）→ raw 下载 URL。上游 `mrdRawUrl`。
pub fn mrd_raw_url(path: &str) -> String {
    format!("{MRD_RAW_BASE}{path}")
}

/// 资源库条目。上游 `RuleResourceCatalogItem`（`ui/src/shared/types/rules.ts`）。
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct RuleResourceCatalogItem {
    pub id: String,
    /// `geosite` / `geoip` / `geosite-lite` / `geoip-lite` / `custom`。
    pub category: String,
    pub name: String,
    /// meta-rules-dat 仓库内路径。
    pub path: String,
    /// 该条目**已随包出厂**（`builtin_geo_rulesets` 表内同名 tag），资源库里应显示为「已内置」而非可下载。
    ///
    /// 字段保留而非删除，尽管本表投影出来的条目恒为 `true`：真正需要它的是**外置 tab** ——
    /// 远端全量清单里也会出现随包同名项，那边必须逐条现算才知道该不该标「已内置」
    /// （`commands/rules.rs` 的 `catalog_item_from_tree_path`）。
    ///
    /// 恒由 Rust 现算、**不从缓存回读**（见 `commands/rules.rs` 的 `parse_cached_catalog_item`）：
    /// 随包内容随版本走，缓存却是上个版本落的盘，信缓存等于让旧版本的随包清单决定新版本的 UI。
    pub bundled: bool,
}

/// catalog 拉取结果。上游 `RuleResourceCatalogResult`。
///
/// `source` **自述来源**（`remote` / `cache` / `builtin`）—— 这是本 DTO 最要紧的字段：
/// 无网络时返回内置精选表并标 `source:"builtin"` 是**诚实降级**（前端可据此提示「离线清单」），
/// 与「谎称拉到了远端全量」是两回事。
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RuleResourceCatalogResult {
    pub items: Vec<RuleResourceCatalogItem>,
    /// 远端拉取时间戳（ms）；内置回退为 `None` → TS `fetchedAt: number | null`。
    pub fetched_at: Option<i64>,
    pub source: String,
}

/// 单条随包 geo → 资源库条目。
///
/// `name` 取自 **tag**、`path` 取自 **file_name**，两者只在 `geosite-category-ai` 这一条上分岔
/// （tag 是 `geosite-category-ai`，上游文件却叫 `category-ai-!cn.srs`）。分岔必须保留：
/// - `name` 若跟着文件名走 → 应用分流弹窗的标签池会出现 `category-ai-!cn`，用它建的预设生成
///   `geosite-category-ai-!cn`，不在随包表内 → fail-closed 剪枝，规则静默失效；
/// - `path` 若跟着 tag 走 → 下载 URL 指向上游不存在的 `geo/geosite/category-ai.srs` → 404。
fn catalog_item_of(b: BuiltinGeoRuleSet) -> RuleResourceCatalogItem {
    let kind = match b.category {
        GeoCategory::Geosite => "geosite",
        GeoCategory::Geoip => "geoip",
    };
    let prefix = format!("{kind}-");
    let name = b.tag.strip_prefix(&prefix).unwrap_or(&b.tag).to_string();
    let stem = b.file_name.strip_suffix(".srs").unwrap_or(&b.file_name);
    let bare = stem.strip_prefix(&prefix).unwrap_or(stem);
    RuleResourceCatalogItem {
        path: format!("geo/{kind}/{bare}.srs"),
        // 投影自随包表 ⇒ 恒真，无须再查一次 `is_bundled_geo_tag`。
        bundled: true,
        id: b.tag.clone(),
        category: kind.to_string(),
        name,
    }
}

/// 内置清单 = 随包清单的投影。
pub fn rule_resource_catalog() -> Vec<RuleResourceCatalogItem> {
    builtin_geo_rulesets()
        .into_iter()
        .map(catalog_item_of)
        .collect()
}

/// 按 id 查条目。上游 `findCatalogItem`。
pub fn find_catalog_item(id: &str) -> Option<RuleResourceCatalogItem> {
    rule_resource_catalog().into_iter().find(|i| i.id == id)
}

/// 内置回退的 catalog 结果（`source:"builtin"`, `fetchedAt:null`）。
pub fn builtin_catalog_result() -> RuleResourceCatalogResult {
    RuleResourceCatalogResult {
        items: rule_resource_catalog(),
        fetched_at: None,
        source: "builtin".to_string(),
    }
}

#[cfg(test)]
mod tests;

File: ui/src/domain/rule-resource-catalog.ts
/**
 * 规则资源库（catalog）—— **前端不再持有清单本体**。
 *
 * 内置清单（28 条 = 随包表 `builtin_geo_rulesets()` 的投影）的单一真值在 Rust
 * （`crates/config-engine/src/user_config/rule_resource_catalog.rs`），经 `rule_resources_get_catalog`
 * / `rule_resources_refresh_catalog` 下发（`RuleResourceCatalogResult`，`source` 字段自述来源）。
 * 依治理文档 §3.1 Q3：**常量表 → Rust SoT + 一次 invoke 拉取入 store**。
 * 迁移当时已逐条对拍取证：33/33 全等（category/name/path 派生），并对 Rust 表变异验证转红
 * ——「33」是当时两侧共有的条数，其后 Rust 侧收敛成随包表的投影（28 条，81a4e68）。
 *
 * `MRD_RAW_BASE` / `mrdRawUrl` / `RULE_RESOURCE_CATALOG` / `findCatalogItem` 一并删除：
 * 下载 URL 的拼装发生在**下载那一侧**（Rust），前端不下载，留着就是第二份基址常量。
 *
 * 本模块只剩 `deriveResourceMeta`（②类 UI 派生：手动 URL 输入时的**即时预填**，同步渲染路径）。
 * 审计 §B 记它为「资源库弹窗」的待接线物料，弹窗未实现故暂无消费方 —— 按 §B 判据保留，勿当死代码删。
 *
 * ⚠️ 已登记漂移敞口：将来 Rust 侧接下载时会需要一份等价的「URL → category/name」兜底命名
 * （原注释即写明「主进程兜底与渲染端预填共用」）。届时应以 Rust 为准、前端谓词退役，
 * **不要两侧各留一份**（那正是本批消灭的形态）。
 */
import type { RuleResourceCategory } from '../contracts/types';

/** 分类显示名：技术类别固定显示 Geosite/GeoIP；custom 文案由调用方按当前 locale 注入。 */
export function categoryLabel(cat: RuleResourceCategory, customLabel: string): string {
  switch (cat) {
    case 'geosite':
    case 'geosite-lite':
      return 'Geosite';
    case 'geoip':
    case 'geoip-lite':
      return 'GeoIP';
    default:
      return customLabel;
  }
}

/**
 * 资源库条目的「已具备」状态：随包出厂 / 已下载 / 都不是（= 真正可下载）。
 *
 * 抽成纯函数而非写在弹窗里：本仓 vitest 是 node 环境无 jsdom，组件交互测不了，判据留在组件里
 * 等于没有门。而这条判据错一次的代价是实打实的 —— 判漏 → 用户对着已在盘上的资源反复下载；
 * 判多 → 用户以为规则集在手，实际生成配置时该 tag 无处可寻（fail-closed 剪枝，规则静默失效）。
 *
 * **随包优先于已下载**（两者同真时返 `'bundled'`）：`route.rs` 生成 rule_set 时先注入随包副本、
 * 再对未定义的 tag 才去查 `config.ruleResources`，即随包在位时**生效的恒是随包那份**，下载副本
 * 只在随包 `.srs` 缺失/损坏时才顶上。此时标「已下载」会让用户以为生效的是自己下的那份。
 */
export type CatalogItemStatus = 'bundled' | 'downloaded' | null;

export function catalogItemStatus(
  item: { id: string; bundled?: boolean },
  downloadedIds: ReadonlySet<string>,
): CatalogItemStatus {
  if (item.bundled) return 'bundled';
  return downloadedIds.has(item.id) ? 'downloaded' : null;
}

/** 资源库条目排序/去重只需要这三个字段（组件传的是完整 `RuleResourceCatalogItem`）。 */
type CatalogSortable = { id: string; name: string; bundled?: boolean };

/**
 * 某个 tab 该显示哪些条目、按什么序 —— 「外置去重」+「已具备优先、名称次之」两件事的唯一落点。
 *
 * ## 去重：外置排掉与内置 id 重合的条目
 *
 * 判据是代码事实不是偏好：`crates/config-engine/src/builder/route.rs` 的 `add_local_geo_rule_set`
 * 在同 id 时**优先注入随包那份**，即外置清单里与内置重合的条目下载了也用不上（纯白下）。
 * 列出来只会让用户以为自己缺了什么。
 *
 * 实测重合面（2026-07-30，盘上缓存 `<userData>/rule-resource/catalog.json` 对随包 28 条）：
 * 外置 2176 条里重合 **27** 条 —— 唯一没重合的随包项是 `geosite-category-ai`（上游文件叫
 * `category-ai-!cn.srs` ⇒ 外置侧 id 是 `geosite-category-ai-!cn`，与随包 tag 不同形，见
 * `builtin_geo_rulesets.rs` 的 `app_geo_entry`）。⇒ 过滤后条数 2176 → 2149，量级上不显著，
 * 但那 27 条恰恰是最常被点的（cn / youtube / telegram / netflix…），去重收益集中在这里。
 *
 * ## 排序：已具备 > 名称
 *
 * ⚠️ 内置 tab 恒 `bundled:true`（`81a4e68` 把它收敛成随包表的投影）⇒ 该 tab 上第一级键恒相等、
 * 只剩名称排序。**这不是排序没生效**，是那次收敛的必然结果；实际因状态键改观的只有外置 tab。
 *
 * 名称同名（`geosite-netflix` / `geoip-netflix` 的 name 都是 `netflix`）时不再加第三级键：
 * `Array.prototype.sort` 自 ES2019 起保证稳定，同名项保持入参序即可，多一级键只是多一处要维护的判据。
 */
export function catalogTabItems<T extends CatalogSortable>(
  tab: 'builtin' | 'external',
  builtin: readonly T[],
  external: readonly T[],
  downloadedIds: ReadonlySet<string>,
): T[] {
  let src: readonly T[] = builtin;
  if (tab === 'external') {
    const builtinIds = new Set(builtin.map((it) => it.id));
    src = external.filter((it) => !builtinIds.has(it.id));
  }
  // 已具备(0) 在前、待下载(1) 在后；同档按名称。两级键缺任一级都会让排序退化成另一半。
  const held = (it: T) => (catalogItemStatus(it, downloadedIds) === null ? 1 : 0);
  return [...src].sort((a, b) => held(a) - held(b) || a.name.localeCompare(b.name));
}

/**
 * 资源库列表区一行都渲染不出来时该说什么 —— 这是一个**解释正确性**的判据，四态各有排他理由。
 *
 * 顺序不是排版偏好，每一条都在挡一句谎话：
 *  1. `count > 0` → `null`：有行就不是空态。**刷新失败但上一份缓存还在时不许拿失败态把清单顶掉**
 *     —— 用户点刷新前看得见的东西，点完反而没了，比不给反馈更糟。那一路的持久解释挂在
 *     外置状态行（`extStatusText`）上，不在这里。
 *  2. `total > 0` → `noMatch`：清单在手、只是被搜索过滤空了。此时报「加载失败」是反过来说谎。
 *  3. `error` → `error`：清单没拿到。这一条必须压过下面的 `noMatch` —— 加载失败时 `count` 同样是 0，
 *     不排在前面的话用户看到「无匹配资源」，而他根本没搜任何东西。**这就是本次要修的那句谎话。**
 *  4. `notFetched` → `notFetched`：外置 tab 一次都没拉过（既没预载到缓存、也没点过刷新）。
 *
 * 抽成纯函数的理由同 `catalogItemStatus`：本仓 vitest 无 jsdom，判据写在 JSX 三元里等于没有门。
 */
export type CatalogEmptyKind = 'error' | 'notFetched' | 'noMatch' | null;

export function catalogEmptyKind(p: {
  /** 当前 tab 的加载失败原因；`null` = 没出错。空串也算出错（后端未必给得出 message）。 */
  error: string | null;
  /** 外置 tab 且一次清单都没拿到。 */
  notFetched: boolean;
  /** 本 tab 去重排序后的总条数（**未经**搜索过滤）。 */
  total: number;
  /** 经搜索过滤后可渲染的行数。 */
  count: number;
}): CatalogEmptyKind {
  if (p.count > 0) return null;
  if (p.total > 0) return 'noMatch';
  if (p.error !== null) return 'error';
  if (p.notFetched) return 'notFetched';
  return 'noMatch';
}

/** 从下载 URL 推导资源 category 与默认 name（手动 URL 自动命名，渲染端预填）。 */
export function deriveResourceMeta(url: string): { category: RuleResourceCategory; name: string } {
  let pathname = url;
  try {
    pathname = new URL(url).pathname;
  } catch {
    pathname = url.split('?')[0].split('#')[0];
  }
  const segs = pathname.split('/').filter(Boolean);
  let base = segs.length ? segs[segs.length - 1] : 'rule';
  try {
    base = decodeURIComponent(base);
  } catch {
    /* keep */
  }
  base = base.replace(/\.srs$/i, '').replace(/\.json$/i, '') || 'rule';

  // meta-rules-dat 路径 → 对应分类（含 -lite），name 复用 basename（id 与内置同 ⇒ 自然去重）
  const m = pathname.match(/\/(geo|geo-lite)\/(geosite|geoip)\//i);
  if (m) {
    const lite = m[1].toLowerCase() === 'geo-lite';
    const kind = m[2].toLowerCase(); // geosite | geoip
    const category = (lite ? `${kind}-lite` : kind) as RuleResourceCategory;
    return { category, name: base };
  }

  // asn → AS<number>
  const asn = pathname.match(/\/asn\/(AS\d+)\.srs$/i);
  if (asn) return { category: 'custom', name: asn[1].toUpperCase() };

  return { category: 'custom', name: base.replace(/[^\w.-]+/g, '_') };
}
